### PR TITLE
Add serializable class for DockerCmdExecFactoryImpl

### PR DESCRIPTION
**Pull request by KostyaSha**
Pair for DockerClientConfig for building DockerClient after deserialization on remote side.

[Review on Reviewable](https://reviewable.io/reviews/docker-java/docker-java/379)

### CODE REVIEW

**Comment by KostyaSha** [comment]
DockerClient need to be created on remote jvm and configuration should be Serializable.
extracting from jenkins docker-plugin:
https://github.com/jenkinsci/docker-plugin/commit/c200189fc6a63a31b01309c02816bcefbb1a9fb4#diff-bccfbf21349af601aaa58e3583b6f8f7R177

https://github.com/jenkinsci/docker-plugin/commit/c200189fc6a63a31b01309c02816bcefbb1a9fb4#diff-5cfb667ff07f3ac66a578a9efcd842e9R171

@marcuslinke for review

File: src/main/java/com/github/dockerjava/jaxrs/DockerCmdExecConfig.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+import java.io.Serializable;
+
+/**
+ * Serializable object that store options for {@link com.github.dockerjava.jaxrs.DockerCmdExecFactoryImpl}
+ *
+ * @author Kanstantsin Shautsou
+ */
+public class DockerCmdExecConfig implements Serializable {
+    private static final long serialVersionUID = 1L;
```
Comment (marcuslinke) [inline src/main/java/com/github/dockerjava/jaxrs/DockerCmdExecConfig.java]: use generated serialVersionUID

Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/jaxrs/DockerCmdExecConfig.java]: Every newly added file should have `1L` generated must be extracted only for already used classes.

File: src/main/java/com/github/dockerjava/jaxrs/DockerCmdExecConfig.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    public Integer getMaxTotalConnections() {
+        return maxTotalConnections;
+    }
+
+    public DockerCmdExecConfig setMaxTotalConnections(Integer maxTotalConnections) {
+        this.maxTotalConnections = maxTotalConnections;
+        return this;
+    }
+}
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/jaxrs/DockerCmdExecConfig.java]: add new line

File: src/main/java/com/github/dockerjava/jaxrs/DockerCmdExecConfig.java (hunk 1 above)
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/jaxrs/DockerCmdExecConfig.java]: https://github.com/jenkinsci/ssh-agent-plugin/blob/master/CONTRIBUTING.md#resulting-from-long-experience

**Comment by KostyaSha** [comment]
@marcuslinke refactored, does it look ok now?

**Comment by marcuslinke** [comment]
Better annotate `@CheckForNull` at getters?

**Comment by KostyaSha** [comment]
Right)

**Comment by KostyaSha** [comment]
@marcuslinke I not sure about duplicating this options, should `DockerCmdExecFactoryImpl` be serializable instead? Or should it store then DockerCmdExecConfig instead options?

**Comment by marcuslinke** [comment]
What's the problem with the current implementation? I don't understand...

**Comment by KostyaSha** [comment]
Duplicates fields and misses other possible fields (not sure for what filters required)

**Comment by marcuslinke** [comment]
As `DockerCmdExecFactoryImpl` is more than a bean I dont want to make it serializable.

Regarding request filters I remember they have been introduced because someone needed basic authentication that was realized this way. Response filters was added because of completeness.

**Comment by KostyaSha** [comment]
> As DockerCmdExecFactoryImpl is more than a bean I dont want to make it serializable.

That what i thought also, would it make sense store ExecConfig in it?

> Regarding request filters I remember they have been introduced because someone needed basic authentication that was realized this way. Response filters was added because of completeness.

Can they be serializable (never used them) to have full serialisable configuration for client?

**Comment by marcuslinke** [comment]
`ClientRequestFilter`s and `ClientResponseFilter`s are customizations that are not guaranteed to be serializable. This raises the question to me if docker-java itself should deal with this topic. Maybe configuration serialization is NOT a library issue? For me it seems the demands are rather determined by the concrete use case. I guess in jenkins for example you don't need any `ClientRequestFilter`s so why should you care about it there? Maybe in other scenarios they are needed, so how should docker-java deal with this different requirements?

**Comment by KostyaSha** [comment]
@marcuslinke jenkins can execute code from master and slave. Execution from slave allows exclude proxying through master. So jenkins slave will connect to docker directly and doesn't cause any load on master. That's why jenkins transfer configs to slave jvm, create client and uses it. docker-java allows serializing part of duality configuration...

**Comment by marcuslinke** [comment]
So if this means excluding `ClientRequestFilter`s and `ClientResponseFilter`s from serialization this is wrong, because docker-java shouldn't developed with jenkins needs in mind. Other use-cases of docker-java may need exactly this serialization.

**Comment by KostyaSha** [comment]
Any valid use case will be valid in jenkins too. Btw jenkins adds ~3k installation usages for this library. 
Filters need to be serialised only if they are containing some data, i think in normal cases Filter class itself can be initialised on remote jvm side. 

My doubt is about replacing timeout and connection values in Factory with `DockerCmdExecConfig` (exclude duplicated storages).

**Comment by KostyaSha** [comment]
Bit related to #594 , not actual now, closing

**Pull request closed without merging**